src/stream.rs

```diff
@@ -159,6 +152,17 @@ impl Stream {
         }
     }
 
+    pub(crate) fn reset(&mut self) -> io::Result<()> {
```

Comment:
I'd somewhat rather have this return `io::Result<Stream>`, and remove the `From<DeadlineStream>`. My thinking is that the `From` is no longer a perfect conversion, since you can wind up with a stream that has leftover deadlines on it. That might be a silent surprise down the line.

Reply:
@jsha Yes in principle I agree, however, this is what I was alluding to with "the rabbit hole".

When we do `Response::into_reader()` we get something like (and the equivalent for content-length case):

```rust
        let stream = DeadlineStream::new(stream, deadline);
...
Box::new(PoolReturnRead::new(unit, ChunkDecoder::new(stream)))
```

`PoolReturnRead(ChunkDecoder(DeadlineStream(Stream)))`

`PoolReturnRead` is defined as:

```rust
impl<R: Read + Sized + Into<Stream>> PoolReturnRead<R> {
    pub fn new(unit: Option<Unit>, reader: R) -> Self {
```

That `Into<Stream>` trait is the rabbit hole, cause when we go:

```rust
let stream: Stream = pool_return_read.into();
```

Rust compiler is smart and unnests all three levels down to `Stream` and in that nesting we got `ChunkDecoder` which is external to ureq.

The good rust solution would be to use the `TryFrom` trait instead of `From` (which `Into` uses). But then we're looking at bigger changes.

Reply:
Ah, that makes sense! I'm convinced this is a good change for now. We can try and make improvements later. Thanks for finding this bug.